File: dialects/Python/rgx.py
#!/usr/bin/env python
#
# -----------------------------------------------------------------------------
# This file is part of rgx.
#
# rgx is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# rgx is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with rgx. If not, see <http://www.gnu.org/licenses/>.
# -----------------------------------------------------------------------------

import re
import json
import argparse
from bottle import error, HTTPError, get, post, request, route, run

@error(400)
def error400(err):
  err.content_type = 'application/json'
  return json.dumps({ 'error': "{0}".format(err.body) })

@get('/flags')
def flags():
  return {
    'i': 'Ignore case',
    'L': 'Locale dependent',
    'm': 'Multi-line',
    's': 'dot matches all',
    'u': 'Unicode dependent'
  }

@post('/')
def test():
  try:
    params = request.json
  except ValueError:
    raise HTTPError(400, "Bad JSON data.")

  print(request.json)
  print()

  ptrn  = request.json['pattern']
  flags = request.json['flags']

  if flags:
    ptrn = '(?' + flags + ')' + ptrn;

  try:
    ptrn  = re.compile(ptrn)
  except Exception as e:
    return {
      'status': 'RC_BADPATTERN',
      'error':  "{0}".format(e)
    }

  subj  = request.json['subject']

  match = ptrn.match(subj)

  if match:
    out = {
      'status': 'RC_MATCH',
      'offset': match.span(),
      'captures': []
    }

    print(out)

    for i in range(1, match.lastindex+1):
      out['captures'].append({
        'position': match.start(i),
        'string': match.group(i)
      })

    print(out)
    return out;
  else:
    return { 'status': 'RC_NOMATCH' }


parser = argparse.ArgumentParser(description='A Python rgx dialect engine.')
parser.add_argument('host', help='the interface to bind to', action='store')
parser.add_argument('port', help='the port to bind to', type=int, action='store')

args = vars(parser.parse_args())

run(host=args['host'], port=args['port'], debug=True)

File: dialects/JavaScript/rgx.js
#!/usr/bin/env js

/**
 * This file is part of rgx.
 *
 * rgx is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * rgx is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with rgx. If not, see <http://www.gnu.org/licenses/>.
 */

/// rgx.js
///
/// rgx::JavaScript engine that allows for executing
/// JavaScript regexes and returns their matches and captures
///
/// Input pattern, subject, and options are expected to be JSON
/// encoded. Output is _always_ JSON encoded.
///
/// Usage: ./rgx.js "pattern" "subject" "options"

var display_usage = function() {
  print('Usage: ./rgx.js "pattern" "subject" "options"');
}

/**
 * Prints an error report due to bad invocation of the script.
 * The script will abort after the error is printed.
 *
 * Output looks like this: { failure: msg }
 */
var invocation_error = function(msg) {
  print(JSON.stringify({ failure: msg }));
  quit(1);
}

if (arguments.length == 1 && (arguments[0] == "--help" || arguments[0] == "-h")) {
  display_usage();
  quit(1);
}
else if (arguments.length < 3) {
  invocation_error("Invalid number of arguments, run with --help for usage.")
  quit(1);
}

print(arguments);

/**
 * Prints an error report due to bad input and aborts the script.
 *
 * The output looks like this: { error: msg }
 */
var input_error = function(cause) {
  print(JSON.stringify({ error: cause }));
  quit(1);
}

var decode = function(arg, name) {
  try {
    return JSON.parse(arg);
  } catch (e) {
    invocation_error("Unable to decode " + name + ": " + e);
    quit(1);
  }
}

var pattern = decode(arguments[0], "pattern"),
    subject = decode(arguments[1], "subject"),
    options = decode(arguments[2], "options"),
    expr    = null;

// Validate the pattern
try {
  expr = RegExp(pattern, options);
} catch (e) {
  input_error(e.message);
}

// Execute it
var result = expr.exec(subject);
if (result) {
  // we got a match, build its range
  var match     = { begin: subject.search(expr), end: null },
      captures  = result.slice(1);

  match.end = match.begin + result[0].length;

  // print the range of the match and the captures as per the spec
  print(JSON.stringify([ match.begin, match.end ].concat(captures) ));
  quit(0);
}
else {
  // no match, print an empty array as per the spec
  print(JSON.stringify( [] ));
  quit(0);
}